File: submissions/tessoku-book/tessoku_book_r.cpp
#include <iostream>
#include <vector>
using namespace std;
int main() {
    int N, S;
    cin >> N >> S;
    vector<int> A(N + 1);
    for (int i = 1; i <= N; i++) {
        cin >> A[i];
    }
    vector<vector<bool>> dp(N + 1, vector<bool>(S + 1));
    // 動的計画法 (i=0)
    dp[0][0] = true;
    for (int i = 1; i <= N; i++) dp[0][i] = false;
    // 動的計画法 (i>=1)
    for (int i = 1; i <= N; i++) {
        for (int j = 0; j <= S; j++) {
            if (j < A[i]) {
                if (dp[i - 1][j] == true)
                    dp[i][j] = true;
                else
                    dp[i][j] = false;
            }
            if (j >= A[i]) {
                if (dp[i - 1][j] == true || dp[i - 1][j - A[i]] == true)
                    dp[i][j] = true;
                else
                    dp[i][j] = false;
            }
        }
    }
    if (dp[N][S] == true)
        cout << "Yes" << endl;
    else
        cout << "No" << endl;
}

File: submissions/tessoku-book/tessoku_book_r.rs
use proconio::{fastout, input};
#[fastout]
fn main() {
    input! {
        n: usize,
        s: usize,
        a: [usize; n],
    }
    match solve(s, &a) {
        true => println!("Yes"),
        false => println!("No"),
    }
}
fn solve(s: usize, a: &Vec<usize>) -> bool {
    let mut dp = vec![vec![false; s + 1]; a.len() + 1];
    dp[0][0] = true;
    for i in 1..dp.len() {
        for j in 0..dp[0].len() {
            if j < a[i - 1] {
                if dp[i - 1][j] == true {
                    dp[i][j] = true;
                }
            }
            if j >= a[i - 1] {
                if dp[i - 1][j] == true || dp[i - 1][j - a[i - 1]] == true {
                    dp[i][j] = true;
                }
            }
        }
    }
    dp[dp.len() - 1][dp[0].len() - 1]
}